### PR TITLE
[optim] Move test_grad_scaling_autocast_fused_optimizers to test_cuda.py 

### PR DESCRIPTION
**Pull request by SandishKumarHN**
this PR address the comments in this PR #124904 

- Move test_grad_scaling_autocast_fused_optimizers to test_cuda.py 
- Combine _grad_scaling_autocast_fused_optimizers into test_grad_scaling_autocast_fused_optimizers
- Move to OptimizerInfo framework. 
- For failing tests test_grad_scaling_autocast_fused_optimizers AdamW_cuda_float32, Adam_cuda_float32
    - Added toleranceOverride in this PR
    - created a issue #127000
  

```
> (c2env) [sandish@devgpu166.ash6 ~/pytorch (refactoroptimizers)]$ python test/test_cuda.py -k test_grad_scaling_autocast_fused_optimizers -v
/home/sandish/pytorch/torch/backends/cudnn/__init__.py:106: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
/home/sandish/pytorch/torch/backends/cudnn/__init__.py:106: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
test_grad_scaling_autocast_fused_optimizers_Adagrad_cpu_float32 (__main__.TestCudaOptimsCPU) ... {'fused': True}
{'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'lr': 0.1, 'fused': True}
{'lr': 0.1, 'fused': True}
{'initial_accumulator_value': 0.1, 'weight_decay': 0.1, 'fused': True}
{'initial_accumulator_value': 0.1, 'weight_decay': 0.1, 'fused': True}
{'lr': 0.1, 'lr_decay': 0.5, 'weight_decay': 0.1, 'fused': True}
{'lr': 0.1, 'lr_decay': 0.5, 'weight_decay': 0.1, 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_AdamW_cpu_float32 (__main__.TestCudaOptimsCPU) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_Adam_cpu_float32 (__main__.TestCudaOptimsCPU) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_SGD_cpu_float32 (__main__.TestCudaOptimsCPU) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
{'momentum': 0.9, 'fused': True}
{'momentum': 0.9, 'fused': True}
{'momentum': 0.9, 'dampening': 0.5, 'fused': True}
{'momentum': 0.9, 'dampening': 0.5, 'fused': True}
{'momentum': 0.9, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'nesterov': True, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'nesterov': True, 'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_Adagrad_cuda_float32 (__main__.TestCudaOptimsCUDA) ... skipped 'cuda is not supported for fused on Adagrad'
test_grad_scaling_autocast_fused_optimizers_AdamW_cuda_float32 (__main__.TestCudaOptimsCUDA) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'capturable': True, 'fused': True}
{'capturable': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'capturable': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'capturable': True, 'fused': True}
{'lr': tensor(0.0010), 'amsgrad': True, 'capturable': True, 'fused': True}
{'lr': tensor(0.0010), 'amsgrad': True, 'capturable': True, 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_Adam_cuda_float32 (__main__.TestCudaOptimsCUDA) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'fused': True}
{'capturable': True, 'fused': True}
{'capturable': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'capturable': True, 'fused': True}
{'weight_decay': 0.1, 'amsgrad': True, 'capturable': True, 'fused': True}
{'lr': tensor(0.0010), 'amsgrad': True, 'capturable': True, 'fused': True}
{'lr': tensor(0.0010), 'amsgrad': True, 'capturable': True, 'fused': True}
ok
test_grad_scaling_autocast_fused_optimizers_SGD_cuda_float32 (__main__.TestCudaOptimsCUDA) ... {'fused': True}
{'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': 0.01, 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
{'lr': tensor(0.0010), 'fused': True}
{'momentum': 0.9, 'fused': True}
{'momentum': 0.9, 'fused': True}
{'momentum': 0.9, 'dampening': 0.5, 'fused': True}
{'momentum': 0.9, 'dampening': 0.5, 'fused': True}
{'momentum': 0.9, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'nesterov': True, 'weight_decay': 0.1, 'fused': True}
{'momentum': 0.9, 'nesterov': True, 'weight_decay': 0.1, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
{'weight_decay': 0.1, 'maximize': True, 'fused': True}
ok

----------------------------------------------------------------------
Ran 8 tests in 16.117s

OK (skipped=1)

> lintrunner test/test_cuda.py
----------------------------------------------------------------------
ok No lint issues.

> lintrunner torch/testing/_internal/common_optimizers.py
----------------------------------------------------------------------
ok No lint issues.
```

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -4702,6 +4703,98 @@ class TestCudaOptims(TestCase):
     # These tests will be instantiate with instantiate_device_type_tests
     # to apply the new OptimizerInfo structure.
 
+    @onlyCUDA
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    @parametrize("foreach, fused", [(False, False), (True, False), (False, True)])
```
Comment (janeyx99) [inline test/test_cuda.py:4710]: why is this necessary?

**Review by janeyx99 — COMMENTED** [review]
All the tests skipping is weird...what's your hardware?

File: test/test_cuda.py [hunk 2]
```diff
@@ -4702,6 +4703,98 @@ class TestCudaOptims(TestCase):
...
+    @optims(
+        [
+            optim
+            for optim in optim_db
+            if "foreach" in optim.supported_impls and "fused" in optim.supported_impls
```
Comment (janeyx99) [inline test/test_cuda.py:4715]:
```suggestion
            if "fused" in optim.supported_impls
```

File: test/test_cuda.py [hunk 3]
```diff
@@ -4702,6 +4703,92 @@ class TestCudaOptims(TestCase):
...
+        if device not in optim_info.supports_fused_on:
+            self.skipTest(
+                f"{device} is not supported for fused on {optim_info.optim_cls.__name__}"
+            )
+        optim_inputs = optim_info.optim_inputs_func(device=device)
+        optim_cls = optim_info.optim_cls
+        for optim_input in optim_inputs:
+            for _separate_unscale in (True, False):
+                kwargs = optim_input.kwargs
+                # kwargs["fused"] = False
```
Comment (SandishKumarHN) [inline test/test_cuda.py:4731]: had to comment this out for below error. 

https://github.com/pytorch/pytorch/blob/main/torch/optim/adagrad.py#L69

```
RuntimeError: `fused=True` requires all the params to be floating point Tensors of supported devices: ['xpu', 'cpu', 'privateuseone'].
```

File: test/test_cuda.py (hunk 1 above)
Comment (SandishKumarHN) [reply test/test_cuda.py]: @janeyx99 I have meta devgpu, it looks like the supports_fused_on didn't had "cuda:0" entry. 

all 4 fused tests are passing (4 fused entries in [optim_db](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_optimizers.py#L1150) )

File: test/test_cuda.py (hunk 3 above)
Comment (janeyx99) [reply test/test_cuda.py:4727]: Well, this is actually the whole point of the test, so we cannot comment it out.

What we should do is remove the onlyCUDA, and still keep this skip:
```
   if device not in optim_info.supports_fused_on:
            self.skipTest(
                f"{device} is not supported for fused on {optim_info.optim_cls.__name__}"
            )
```

Comment (SandishKumarHN) [reply test/test_cuda.py]: @janeyx99  that worked, but one test is failing at https://github.com/pytorch/pytorch/blob/main/torch/optim/adagrad.py#L540 with 
```   File "/home/sandish/pytorch/torch/optim/adagrad.py", line 540, in _fused_adagrad
    torch._fused_adagrad_(
  File "/home/sandish/pytorch/torch/__init__.py", line 2023, in __getattr__
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
AttributeError: module 'torch' has no attribute '_fused_adagrad_'
```

Comment (janeyx99) [reply test/test_cuda.py]: that might be ur local build being outdated

Comment (janeyx99) [reply test/test_cuda.py]: there is a lint failure though, please run lintrunner before committing

File: test/test_cuda.py [hunk 4]
```diff
@@ -4702,6 +4703,89 @@ class TestCudaOptims(TestCase):
     # These tests will be instantiate with instantiate_device_type_tests
     # to apply the new OptimizerInfo structure.
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
```
Comment (janeyx99) [inline test/test_cuda.py:4708]:
```suggestion
    @onlyNativeDeviceTypes
```

why are we adding this skip?

You should be able to just copy the test case exactly...

File: test/test_cuda.py [hunk 5]
```diff
@@ -4702,6 +4703,89 @@ class TestCudaOptims(TestCase):
...
+    @optims(
+        [optim for optim in optim_db if "fused" in optim.supported_impls],
+        dtypes=[torch.float32],
+    )
+    def test_grad_scaling_autocast_fused_optimizers(self, device, dtype, optim_info):
+        # This ut is from test_cuda.py test_grad_scaling_autocast_fused_optimizers
+        # but only test Adam/AdamW on CPU
+        # TODO: haozhe, support SGD and unified this ut with the CUDA only one
```
Comment (janeyx99) [inline test/test_cuda.py:4716]:
```suggestion
```

**Review by janeyx99 — COMMENTED** [review]
Please copy the test over without any changes and run the test locally with the verbose flag. 

Could you also add a temporary print statement to line after 4727:
```
print(kwargs)
```
so we know what gets run?

File: test/test_cuda.py [hunk 6]
```diff
@@ -4702,6 +4703,89 @@ class TestCudaOptims(TestCase):
...
+                (
+                    mod_control,
+                    mod_scaling,
+                    opt_control,
+                    opt_scaling,
+                    data,
+                    loss_fn,
+                    _,
+                ) = _create_scaling_case(
+                    optimizer_ctor=optim_cls, optimizer_kwargs=kwargs, device="cpu"
```
Comment (janeyx99) [inline test/test_cuda.py:4737]:
```suggestion
                    optimizer_ctor=optim_cls, optimizer_kwargs=kwargs, device=device
```

File: torch/testing/_internal/common_optimizers.py [hunk 7]
```diff
@@ -1147,7 +1147,7 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
         optim_inputs_func=optim_inputs_func_adagrad,
         optim_error_inputs_func=optim_error_inputs_func_adagrad,
         supported_impls=("foreach", "differentiable", "fused"),
-        supports_fused_on=("cpu",),
+        supports_fused_on=("cpu"),
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1150]: this change looks wrong

Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1150]: and should be unnecessary

File: test/test_cuda.py (hunk 3 above)
Comment (SandishKumarHN) [reply test/test_cuda.py]: the local lintrunner shows opposite of the PR linterunner. 

```
    You can run `lintrunner -a` to apply this patch.

    964  964 |         # SparseAdam raises a warning and not an error for the first entry. We
    965  965 |         # update it here:
    966  966 |         error_inputs[0].error_type = FutureWarning
    966      |-        error_inputs[
    967      |-            0
    968      |-        ].error_regex = "Passing in a raw Tensor as ``params`` to SparseAdam"
         967 |+        error_inputs[0].error_regex = (
         968 |+            "Passing in a raw Tensor as ``params`` to SparseAdam"
         969 |+        )
    970  970 | 
    971  971 |         error_inputs += [
    972  972 |             ErrorOptimizerInput(
```
but fixed as per PR lintrunner

File: test/test_cuda.py [hunk 8]
```diff
@@ -4702,6 +4704,84 @@ class TestCudaOptims(TestCase):
     # These tests will be instantiate with instantiate_device_type_tests
     # to apply the new OptimizerInfo structure.
 
+    @onlyNativeDeviceTypes
+    @optims(
+        [optim for optim in optim_db if "fused" in optim.supported_impls],
+        dtypes=[torch.float32],
+    )
+    def test_grad_scaling_autocast_fused_optimizers(self, device, dtype, optim_info):
+        if device not in optim_info.supports_fused_on:
```
Comment (janeyx99) [inline test/test_cuda.py:4713]:
```suggestion
        if device.type not in optim_info.supports_fused_on:
```
AdamW, Adam and SGD should not get skipped for "cuda:0" not being in the list. This should really use device.type.

Comment (SandishKumarHN) [reply test/test_cuda.py:4714]: since device is coming as str, doing ```device = device.split(":")[0]```

File: torch/testing/_internal/common_optimizers.py [hunk 9]
```diff
@@ -1255,6 +1255,21 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "TestOptimRenewed",
                 "test_fused_matches_forloop",
             ),
+            DecorateInfo(
+                # Note on tolerances:
+                # test_grad_scaling_autocast_fused_optimizers_Adam_cuda_float32
+                # Mismatched elements: 64 / 64 (100.0%)
+                # Greatest absolute difference: 2.3484230041503906e-05 at index (7, 5) (up to 1e-05 allowed)
+                # Greatest relative difference: 3.567120074876584e-05 at index (5, 5) (up to 1.3e-06 allowed
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1263]: well, we don’t know why this is the case. could you simplify this comment to just link to the issue tracking the discrepancy?

same for below

**Review by janeyx99 — APPROVED** [review]

File: torch/testing/_internal/common_optimizers.py
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1]: If the original tolerances work, pls remove the code in this file as they're not necessary then. Unless the atol override is still needed.

Comment (SandishKumarHN) [reply torch/testing/_internal/common_optimizers.py:1]: atol override is still needed, otherwise the tests fail.
